### PR TITLE
Add  documentation for the  neqo-http3  code

### PR DESCRIPTION
**Pull request by ddragana**
This is based on #1361

### CODE REVIEW

File: neqo-http3/src/connection.rs [hunk 1]
```diff
@@ -59,6 +59,19 @@ enum Http3RemoteSettingsState {
     ZeroRtt(HSettings),
 }
 
+/// States:
+/// - Initializing: this is the state during the QUIC handshake,
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:63]:
```suggestion
/// - `Initializing`: this is the state during the QUIC handshake,
```

File: neqo-http3/src/connection.rs [hunk 2]
```diff
@@ -59,6 +59,19 @@ enum Http3RemoteSettingsState {
...
+/// - ZeroRtt: this is the state during the ZeroRtt phase,
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:64]:
```suggestion
/// - `ZeroRtt`: 0-RTT has been enabled and is active
```

File: neqo-http3/src/connection.rs [hunk 3]
```diff
@@ -59,6 +59,19 @@ enum Http3RemoteSettingsState {
...
+/// - Connected
+/// - GoingAway(StreamId): The connection has received a `GOAWAY` frame
+/// - Closing(ConnectionError): The connection is closed. The closing has been initiated by this
+///   end of the connection, e.g. the `CONNECTION_CLOSE` frame has been sent. In this state, the
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:68]:
```suggestion
///   end of the connection, e.g., the `CONNECTION_CLOSE` frame has been sent. In this state, the
```

File: neqo-http3/src/connection.rs [hunk 4]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
     }
 }
 
+/**
+# HTTP/3 core implementation
+
+This is the core implementation of HTTP/3 protocol. It implements most of the features of the
+protocol. Http3Client and Http3ServerHandler implement only client and server side bbehavior.
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:99]:
```suggestion
protocol. `Http3Client` and `Http3ServerHandler` implement only client and server side behavior.
```

File: neqo-http3/src/connection.rs [hunk 5]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+- functions that correspond to the Http3Client Http3ServerHandler API:
+  - `new`
+  - `close`
+  - `fetch` -  only used by the client-side implementation
+  - `read_data`
+  - `stream_reset_send`
+  - `stream_stop_sending`
+  - `cancel_fetch`
+  - `stream_close_send`
+  - `webtransport_create_session` -  only used by the client-side implementation
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:111]:
```suggestion
- functions that correspond to [WebTransport](https://w3c.github.io/webtransport/) functions:
  - `webtransport_create_session` -  only used by the client-side implementation
```

File: neqo-http3/src/connection.rs [hunk 6]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+  - `queue_control_frame`
+  - `queue_update_priority`
+  - `set_0rtt_settings`
+  - `get_settings`
+  - `state`
+  - `webtransport_enabled`
+
+## Streams
+
+Http3Connection holds a list of stream handlers. Each send and receive-handler is registered in
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:144]:
```suggestion
Each `Http3Connection` holds a list of stream handlers. Each send and receive-handler is registered in
```

File: neqo-http3/src/connection.rs [hunk 7]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+can be closed and removed ane second may still be active.
+
+The only streams that are not registered are the local control stream, local QPACK decoder stream,
+and local QPACK encoder stream. These streams are send-streams and sending data on this stream is
+handled a bit differently. This is done in the `process_sending` function, i.e. the control data is
+sent first and QPACK data is sent after regular stream data is sent because this stream may have
+new data only after regular streams are handled (TODO we may improve this a bit to send QPACK
+commands before headers.)
+
+There are the following types of streams:
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:156]: Are these types of streams or the types of the new stream events?

File: neqo-http3/src/connection.rs [hunk 8]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+- `Encoder`: there is only a receiver stream of this type and the handler is `EncoderRecvStream`.
+- `NewStream`: there is only a receiver stream of this type and the handler is
+               `NewStreamHeadReader`.
+- `Http`: `SendMessage` and `RecvMessage` handlers are responsible for this type of streams.
+- `Push`: `RecvMessage` is responsible for this type of streams.
+- `ExtendedConnect`: `WebTransportSession` is responsible sender and receiver handler.
+- `WebTransport(StreamId)`: `WebTransportSendStream` and `WebTransportRecvStream` are responsible
+                            sender and receiver handler.
+- `Unknown`: These are all other stream types that are not unknown to the current implementation
+             and should be handled properly by the spec, e.g in our implementation the streams are
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:168]:
```suggestion
             and should be handled properly by the spec, e.g., in our implementation the streams are
```

File: neqo-http3/src/connection.rs [hunk 9]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+             reset.
+
+The streams are registered in `send_streams` and `recv_streams` in following ways depending if they
+are local or remote:
+- local streams:
+  - all local stream will ve register with the appropriate handled.
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:174]:
```suggestion
  - all local stream will be registered with the appropriate handler.
```

File: neqo-http3/src/connection.rs [hunk 10]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+- remote streams:
+  - all new incoming streams are registed with NewStreamHeadReader. This is triggered by
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:176]:
```suggestion
  - all new incoming streams are registered with NewStreamHeadReader. This is triggered by
```

File: neqo-http3/src/connection.rs [hunk 11]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+holds a list of its active streams and clean up is done in `remove_extended_connect`.
+
+###  `WebTransportSendStream` and `WebTransportRecvStream`
+
+WebTransport streams are associated with a session. `WebTransportSendStream` and
+`WebTransportRecvStream` hold a reference to the session and are registered in the session. They
+are added to the session upon createion by `Http3Connection`. They will remove unregistered if they
+are reset or canceled.
+
+The call to function `receive` may produced `Http3ClientEvent::DataReadable`. Actual reading of
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:279]:
```suggestion
The call to function `receive` may produce `Http3ClientEvent::DataReadable`. Actual reading of
```

File: neqo-http3/src/connection.rs [hunk 12]
```diff
@@ -79,6 +92,193 @@ impl Http3State {
...
+
+`WebTransportSendStream` and `WebTransportRecvStream` are associated with a `WebTransportSession`
+and they will be canceled if the session is closed. To be avle to do this `WebTransportSession`
+holds a list of its active streams and clean up is done in `remove_extended_connect`.
+
+###  `WebTransportSendStream` and `WebTransportRecvStream`
+
+WebTransport streams are associated with a session. `WebTransportSendStream` and
+`WebTransportRecvStream` hold a reference to the session and are registered in the session. They
+are added to the session upon createion by `Http3Connection`. They will remove unregistered if they
```
Comment (martinthomson) [inline neqo-http3/src/connection.rs:276]: unregistered what? sessions?

File: neqo-http3/src/connection.rs (hunk 7 above)
Comment (ddragana) [reply neqo-http3/src/connection.rs:157]: They are types of streams. 
For example calling the WebTransportSession::stream_type() function will return  Http3StreamType::ExtendedConnect.

Http3Connection needs to handle some stream types differently therefore it needs to know the type of receive- and send-stream-handlers.

File: neqo-http3/src/connection.rs (hunk 12 above)
Comment (ddragana) [reply neqo-http3/src/connection.rs:277]: I will update the sentence.
"The streams are unregistered from the session..."

File: neqo-http3/src/connection.rs [hunk 13]
```diff
@@ -79,6 +92,194 @@ impl Http3State {
...
+are reset or canceled.
```
Comment (ddragana) [inline neqo-http3/src/connection.rs:278]:
```suggestion
`WebTransportRecvStream` hold a reference to the session and are registered in the session upon
  creation by `Http3Connection`. The `WebTransportSendStream` and `WebTransportRecvStream`
  handlers will be unregistered from the session if they are closed, reset or canceled.
```

File: neqo-transport/src/connection/mod.rs [hunk 14]
```diff
@@ -607,13 +607,13 @@ impl Connection {
...
-    /// Get a resumption token.  The correct way to obtain a resumption token is
-    /// waiting for the `ConnectionEvent::ResumptionToken` event.  However, some
-    /// servers don't send `NEW_TOKEN` frames and so that event might be slow in
-    /// arriving.  This is especially a problem for short-lived connections, where
-    /// the connection is closed before any events are released.  This retrieves
-    /// the token, without waiting for the `NEW_TOKEN` frame to arrive.
-    ///
+    /// The correct way to obtain a resumption token is to wait for the
+    /// `ConnectionEvent::ResumptionToken` event. To emit the event we are waiting for a
+    /// resumtion token and a `NEW_TOKEN` frame to arrive. Some servers don't send `NEW_TOKEN`
```
Comment (acreskeyMoz) [inline neqo-transport/src/connection/mod.rs:612]: typo: `resumtion` -> `resumption`

File: neqo-http3/src/connection.rs [hunk 15]
```diff
@@ -126,10 +327,15 @@ impl Http3Connection {
         }
     }
 
+    /// Ths function is called when a not default feature needs to be negotiated. This is currently
```
Comment (acreskeyMoz) [inline neqo-http3/src/connection.rs:330]: Typo: `Ths`

**Review by acreskeyMoz — COMMENTED** [review]
Just found some typos.
The documention is appreciated!

**Pull request merged**